### ISSUE TITLE
Mac: Paste on mouse middle button click

### ISSUE THREAD

**Issue by upachler** [warpdotdev/Warp#416]
### Describe the solution you'd like?

```shell
I use the MacOS terminal quite a lot. One feature I use often and that I miss in Warp is middle mouse button insertion (this is a feature inherited from X Window back in the day).

Basically, you select text in the terminal with the mouse, and then later, in some other context, press the middle mouse button, which causes that text to be inserted where the cursor is. Maybe that doesn't translate 100% to what Warp wants to do; I could imagine that Warp may also want to place the insertion cursor where the user clicked (this is not done by the MacOS terminal).
```

**Comment by rlinq** [comment]
This is also a feature I'm missing in Warp, however I'm used to copy on select and paste with right mouse click. This is super efficient when working on putty terminals on remote x servers from a Windows host. Recently switched to Mac, this is the feature I miss the most. When warp introduced copy on select I was super happy, I can just hope that paste with mouse could be implemented. 

Maybe it's a good idea to make it a config option to choose, middle or right mouse button to paste content.

All the best

**Comment by mkartashev** [comment]
FWIW, this is available in iTerm2 and I also use middle-click to paste a lot. Quite convenient and middle-click isn't used for anything else (not yet anyway).

**Comment by PRNDA** [comment]
+1 for this awesome feature

**Comment by elviskahoro** [comment]
Will reach back out to the team on this.

**Comment by alexander-beedie** [comment]
This definitely needs to exist ;)

**Comment by ncri** [comment]
Yep, missing this a lot as well. One important thing to note: Selecting and pasting something in this way does NOT overwrite what you have currently copied/cut in your clipboard. It will simply insert the selected text directly at the prompt.

**Comment by cdcasey** [comment]
YA +1. Would love to have this.

**Comment by dannyneira** [comment]
Not sure why this was closed, but we're not tracking middle-click paste support here: https://github.com/warpdotdev/Warp/issues/4329

**Comment by dannyneira** [comment]
Duplicate of https://github.com/warpdotdev/Warp/issues/4329

**Comment by joey-squid** [comment]
s/not/now/ I assume :)

**Comment by joey-squid** [comment]
Technically this feature request was for Mac and the other one is for Linux, though it probably makes sense to have one across the board. Would it make sense to retitle/retag the other one so it's not platform-specific?

**Comment by dannyneira** [comment]
Hey Folks, we've released middle-click paste support in today's release [v0.2024.03.12.08.02](https://docs.warp.dev/getting-started/changelog#id-2024.03.14-v0.2024.03.12.08.02).

**Comment by alokedesai** [comment]
Keeping this open to track middle click paste support on Mac

**Comment by jeremyronking** [comment]
> Keeping this open to track middle click paste support on Mac

Appreciate you!

**Comment by wandoschneider-liber** [comment]
Please please please :D

**Comment by jeremyronking** [comment]
Is there an ETA for this feature?  It was there for a hot minute before it was peeled out for Linux only.

**Comment by Punt13140** [comment]
A coworker just said to me "For me, it's a no-go if I can't use middle button". It's the only feature missing for him to move to Warp.

**Comment by PRNDA** [comment]
THREE YEARS!!! 
Is this tooooo hard to implement, or the team just don't want to do it?

**Comment by jabalsad** [comment]
I don't understand why this works on Linux but not Mac OS. Middle-click paste is standard across all terminals.

**Comment by acarl005** [comment]
Hey all! Warp engineer here. First, I'd like to share why this hasn't been done yet. Middle-click paste is very much a concept from the Linux world. Linux actually has separate, independent clipboards. The clipboard that you interact with when you ctrl-c/ctrl-v is separate from the one you interact with when you select/middle-click. Both can have different content concurrently, and Linux desktop environments will provide distinct slots for these.

Of course, several terminal apps have emulated this interaction on other OSes, such that its becoming standard. The complication is that MacOS just doesn't have separate clipboards. It's only got the one. So, the OS isn't providing the APIs to help back this functionality. Basically, Warp has to build its own space for the middle-click clipboard, and that is effectively mutable global state (not trivial in Rust). Middle-click paste also won't work _across apps_ like it does in Linux. For example, on Linux you can select text (without copying with ctrl-c) in Firefox and middle-click paste it into your terminal. This won't be possible on MacOS. We were hesitant to introduce a feature that didn't behave as expected.

That said, the demand for this feature is undeniable. I'm finally looking into building a version of this in Warp. Stay tuned!

**Comment by acarl005** [comment]
Closing as completed as we've shipped this feature in `v0.2024.10.01.08.02.stable_00`. if anyone needs to disable this for some reason, it can be done in preferences > features > editor > middle-clock to paste.
